File: Yandex/Practicum/Algorightms_and_Data_Structures/5_sprint/B._Сбалансированное_дерево/solution.py
import os

LOCAL = os.environ.get('REMOTE_JUDGE', 'false') != 'true'

if LOCAL:
    class Node:
        def __init__(self, value, left=None, right=None):
            self.value = value
            self.right = right
            self.left = left


def dfs(root) -> int:
    if not root:
        return 0
    left_depth: int = dfs(root.left)
    right_depth: int = dfs(root.right)
    if left_depth == float('inf') or right_depth == float('inf') or abs(left_depth - right_depth) > 1:
        return float('inf')
    return max(left_depth, right_depth) + 1


def solution(root) -> bool:
    """
    Time Complexity O(N)
    Memory Complexity O(H)
    """
    return dfs(root) != float('inf')


def test():
    node1 = Node(1)
    node2 = Node(-5)
    node3 = Node(3, node1, node2)
    node4 = Node(10)
    node5 = Node(2, node3, node4)
    assert solution(node5)


if __name__ == '__main__':
    test()

File: Yandex/Practicum/Algorightms_and_Data_Structures/5_sprint/B._Сбалансированное_дерево/solution.cpp
#ifndef REMOTE_JUDGE
struct Node {
  int value;
  const Node* left = nullptr;
  const Node* right = nullptr;
  Node(int value, Node* left, Node* right) : value(value), left(left), right(right) {}
};
#endif
#ifdef REMOTE_JUDGE
#include "solution_tree.h"
#endif
#include <cmath>
#include <iostream>
#include <cassert>
#include <cstdint>
using namespace std;


int dfs(const Node* root) {
    if (!root) {
        return 0;
    }
    int leftDepth = dfs(root->left);
    int rightDepth = dfs(root->right);
    if (leftDepth == INT32_MIN || rightDepth == INT32_MIN || std::abs(leftDepth - rightDepth) > 1) {
        return INT32_MIN;
    }
    return std::max(leftDepth, rightDepth) + 1;
}

bool Solution(const Node* root) {
    /*
    Time Complexity O(N)
    Memory Complexity O(H)
    */
    return dfs(root) != INT32_MIN;
}


#ifndef REMOTE_JUDGE
void test() {
    Node node1({1, nullptr, nullptr});
    Node node2({-5, nullptr, nullptr});
    Node node3({3, &node1, &node2});
    Node node4({10, nullptr, nullptr});
    Node node5({2, &node3, &node4});
    assert(Solution(&node5));
}

int main() {
  test();
}
#endif